### PR TITLE
Infograph 

### PR DESCRIPTION
**Pull request by tonydavis629**
# Pull Request Template

## Description

This PR implements Infograph as a ModularTorchModel. Infograph is a semi-supervised graph convolutional neural network.

## Type of change

Please check the option that is related to your PR.

- [x] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [x] Breaking change (fix or feature that would cause existing functionality to not work as expected)
  - In this case, we recommend to discuss your modification on GitHub issues before creating the PR
- [x] Documentations (modification for documents)

## Checklist

- [x] My code follows [the style guidelines of this project](https://deepchem.readthedocs.io/en/latest/development_guide/coding.html)
  - [x] Run `yapf -i <modified file>` and check no errors (**yapf version must be  0.32.0**)
  - [x] Run `mypy -p deepchem` and check no errors
  - [x] Run `flake8 <modified file> --count` and check no errors
  - [x] Run `python -m doctest <modified file>` and check no errors
- [x] I have performed a self-review of my own code
- [x] I have commented my code, particularly in hard-to-understand areas
- [x] I have made corresponding changes to the documentation
- [x] I have added tests that prove my fix is effective or that my feature works
- [x] New unit tests pass locally with my changes
- [x] I have checked my code and corrected any misspellings

### CODE REVIEW

File: deepchem/models/tests/test_infograph.py [hunk 1]
```diff
@@ -0,0 +1,114 @@
...
+@pytest.mark.torch
+def get_dataset(mode='classification', num_tasks=1):
+    np.random.seed(123)
+    data_points = 20
+    featurizer = MolGraphConvFeaturizer(use_edges=True)
+
+    if mode == 'classification':
+        tasks, all_dataset, transformers = load_bace_classification(featurizer)
+    else:
+        tasks, all_dataset, transformers = load_delaney(featurizer)
```
Comment (rbharath) [inline deepchem/models/tests/test_infograph.py:23]: This will load the full dataset which may be slow. Perhaps better to make a tiny dataset of 20 compounds directly for testing (the data_points specification below only avoids training on the full set I think and doesn't avoid the load)

File: deepchem/models/torch_models/infograph.py [hunk 2]
```diff
@@ -0,0 +1,435 @@
+from deepchem.models.torch_models.modular import ModularTorchModel
+import torch
+import torch.nn as nn
+import torch.nn.functional as F
+from torch.nn import Sequential, Linear, ReLU, GRU
+from deepchem.feat.graph_data import BatchGraphData
+import math
+
+
+class Encoder(torch.nn.Module):
```
Comment (rbharath) [inline deepchem/models/torch_models/infograph.py:10]: Perhaps InfographEncoder since we will have many encoder layers

File: deepchem/models/torch_models/infograph.py [hunk 3]
```diff
@@ -0,0 +1,435 @@
...
+            m = F.relu(self.conv(out, data.edge_index, data.edge_features))
+            out, h = self.gru(m.unsqueeze(0), h)
+            out = out.squeeze(0)
+            feat_map = out
+
+        out = self.set2set(out, data.graph_index)
+        return out, feat_map
+
+
+class FF(nn.Module):
```
Comment (rbharath) [inline deepchem/models/torch_models/infograph.py:53]: Can we reuse one of the existing feed forward nn.modules? I think we have a few

File: deepchem/models/torch_models/infograph.py [hunk 4]
```diff
@@ -0,0 +1,435 @@
...
+    edge_features: int
+        Number of edge features for each input
+    dim: int
+        Dimension of the embedding
+    use_unsup_loss: bool
+        Whether to use the unsupervised loss
+    separate_encoder: bool
+        Whether to use a separate encoder for the unsupervised loss
+
+    Examples
```
Comment (rbharath) [inline deepchem/models/torch_models/infograph.py:103]: I think this is only the fine tuning use case? Can you add an example for the semisupervised pretraining?

File: deepchem/models/torch_models/infograph.py [hunk 5]
```diff
@@ -0,0 +1,435 @@
...
+            Average the result over samples.
+
+        Returns:
+        -------
+        Ep: torch.Tensor
+            Positive part of the divergence / difference.
+        """
+        log_2 = math.log(2.)
+
+        if self.measure == 'GAN':
```
Comment (rbharath) [inline deepchem/models/torch_models/infograph.py:302]: All these different values for measure should be documented in the doctoring for the class

File: docs/source/api_reference/models.rst [hunk 6]
```diff
@@ -393,6 +393,11 @@ Note that this is an alternative implementation for MPNN and currently you can o
 .. autoclass:: deepchem.models.torch_models.MPNNModel
   :members:
 
+InfoGraphModel
```
Comment (rbharath) [inline docs/source/api_reference/models.rst:396]: All the layers should be in the docs. We're trying to build up and document DeepChem's collection of layers so users can start to reuse layers

File: deepchem/models/tests/test_infograph.py (hunk 1 above)
Comment (tonydavis629) [reply deepchem/models/tests/test_infograph.py:23]: I can change it, but these functions are used in about a dozen tests. Maybe would be better to a create a load function specifically for testing.

File: deepchem/models/torch_models/infograph.py (hunk 3 above)
Comment (tonydavis629) [reply deepchem/models/torch_models/infograph.py:53]: It is a feedforward with a skip connection, which does not match any of the current FF layers

File: deepchem/models/torch_models/infograph.py (hunk 4 above)
Comment (tonydavis629) [reply deepchem/models/torch_models/infograph.py:103]: This is the semi-supervised training. There is no infograph training loop that is strictly supervised. There is only semi-supervised and unsupervised.

File: deepchem/models/torch_models/infograph.py [hunk 7]
```diff
@@ -0,0 +1,441 @@
...
+            m = F.relu(self.conv(out, data.edge_index, data.edge_features))
+            out, h = self.gru(m.unsqueeze(0), h)
+            out = out.squeeze(0)
+            feat_map = out
+
+        out = self.set2set(out, data.graph_index)
+        return out, feat_map
+
+
+class FF_skip(nn.Module):
```
Comment (rbharath) [inline deepchem/models/torch_models/infograph.py:53]: Minor note, FFSkip would be better since we don't use snake case in class names

File: deepchem/models/torch_models/infograph.py (hunk 4 above)
Comment (rbharath) [reply deepchem/models/torch_models/infograph.py:108]: Ah I see. Can you add an unsupervised example as well and add a discussion to the doctoring about the two modes of training and how they differ? Lots of detail is good here since this is the first model of its type in deepchem

File: docs/source/api_reference/models.rst [hunk 8]
```diff
@@ -393,6 +393,21 @@ Note that this is an alternative implementation for MPNN and currently you can o
...
+--------------
+.. autoclass:: deepchem.models.torch_models.InfoGraphModel
+  :members:
+
+InfoGraphEncoder
+----------------
```
Comment (rbharath) [inline docs/source/api_reference/models.rst:402]: We want these to go with the other layers since they aren't models and instead are nn.layers

File: deepchem/models/tests/test_infograph.py [hunk 9]
```diff
@@ -0,0 +1,103 @@
+import os
+import pytest
+import numpy as np
+import deepchem as dc
+from deepchem.feat.molecule_featurizers import MolGraphConvFeaturizer
+
+
+@pytest.mark.torch
+def get_dataset(mode='classification'):
```
Comment (rbharath) [inline deepchem/models/tests/test_infograph.py:9]: Let's split this up into two simpler helper functions, one for classification and one for regression? A good rule of thumb is that a unit test shouldn't take thought to read through. This helper is complex enough it requires additional thought

File: deepchem/models/torch_models/infograph.py [hunk 10]
```diff
@@ -0,0 +1,429 @@
...
+        self.lin0 = torch.nn.Linear(num_features, dim)
+
+        nn = Sequential(Linear(edge_features, 128), ReLU(),
+                        Linear(128, dim * dim))
+        self.conv = NNConv(dim, dim, nn, aggr='mean', root_weight=False)
+        self.gru = GRU(dim, dim)
+
+        self.set2set = Set2Set(dim, processing_steps=3)
+
+    def forward(self, data):
```
Comment (rbharath) [inline deepchem/models/torch_models/infograph.py:40]: Can you add a docstring here and document the expected shape of `data`?

File: deepchem/models/tests/test_infograph.py [hunk 11]
```diff
@@ -0,0 +1,103 @@
...
+                           use_unsup_loss=False,
+                           separate_encoder=False)
+
+    model.fit(dataset, nb_epoch=100)
+    scores = model.evaluate(dataset, [metric])
+    assert scores['mean-roc_auc_score'] >= 0.9
+
+
+@pytest.mark.torch
+def test_fit_restore():
```
Comment (rbharath) [inline deepchem/models/tests/test_infograph.py:78]: Can we add a unit test for the unsupervised pretraining?

File: deepchem/models/torch_models/infograph.py [hunk 12]
```diff
@@ -0,0 +1,429 @@
...
+
+    def __init__(self,
+                 num_features,
+                 edge_features,
+                 dim,
+                 use_unsup_loss=False,
+                 separate_encoder=False,
+                 measure='JSD',
+                 average_loss=True,
+                 **kwargs):
```
Comment (rbharath) [inline deepchem/models/torch_models/infograph.py:116]: Do we need `kwargs`? In general better not to use `kwargs` since that introduces a bigger possibility for bugs

File: deepchem/models/torch_models/infograph.py [hunk 13]
```diff
@@ -0,0 +1,429 @@
...
+            'global_d':
+                MultilayerPerceptron(2 * self.embedding_dim,
+                                     self.embedding_dim, (self.embedding_dim,),
+                                     skip_connection=True),
+        }
+
+    def build_model(self):
+        return InfoGraph(**self.components)
+
+    def loss_func(self, inputs, labels, weights):
```
Comment (rbharath) [inline deepchem/models/torch_models/infograph.py:161]: This should have a docstring

File: deepchem/models/torch_models/infograph.py [hunk 14]
```diff
@@ -0,0 +1,429 @@
...
+                unsup_sup_loss = self.unsup_sup_loss(inputs, labels, weights)
+                loss = sup_loss + unsup_loss + unsup_sup_loss * self.learning_rate
+            else:
+                loss = sup_loss + unsup_loss * self.learning_rate
+            return (loss * weights).mean()
+        else:
+            sup_loss = F.mse_loss(self.model(inputs), labels)
+            return (sup_loss * weights).mean()
+
+    def unsup_loss(self, inputs):
```
Comment (rbharath) [inline deepchem/models/torch_models/infograph.py:175]: I think this is a private helper so should be `_unsup_loss`

File: deepchem/models/torch_models/infograph.py [hunk 15]
```diff
@@ -0,0 +1,429 @@
...
+    >>> edge_dim = max([ds.X[i].num_edge_features for i in range(len(ds))])
+    >>> model = InfoGraphModel(num_feat, edge_dim, 15, use_unsup_loss=True, separate_encoder=True)
+    >>> loss = model.fit(ds, nb_epoch=1)
+    """
+
+    def __init__(self,
+                 num_features,
+                 edge_features,
+                 dim,
+                 use_unsup_loss=False,
```
Comment (rbharath) [inline deepchem/models/torch_models/infograph.py:112]: I'm feeling so-so about the use of a flag to gate the unsupervised loss. Do we want to establish that as the defacto pattern for how self-supervised models turn pretraining on/off? Let's discuss offline before we make the final call

File: deepchem/models/torch_models/infograph.py [hunk 16]
```diff
@@ -0,0 +1,429 @@
...
+        y, M = self.components['encoder'](inputs)
+        y_, M_ = self.components['unsup_encoder'](inputs)
+
+        g_enc = self.components['ff1'](y)
+        g_enc1 = self.components['ff2'](y_)
+
+        loss = self.global_global_loss_(g_enc, g_enc1)
+        return loss
+
+    def local_global_loss_(self, l_enc, g_enc, batch):
```
Comment (rbharath) [inline deepchem/models/torch_models/infograph.py:217]: Do you want to do `_local_global_loss` to mark this as private? I think the underscore should come up front

File: deepchem/models/torch_models/infograph.py [hunk 17]
```diff
@@ -0,0 +1,429 @@
...
+
+        if self.average_loss:
+            return Eq.mean()
+        else:
+            return Eq
+
+
+class InfoGraph(torch.nn.Module):
+    """
+    The nn.Module for InfoGraph. This class defines the forward pass of InfoGraph.
```
Comment (rbharath) [inline deepchem/models/torch_models/infograph.py:365]: Can you add a usage example?

File: deepchem/models/torch_models/layers.py [hunk 18]
```diff
@@ -25,68 +25,69 @@ class MultilayerPerceptron(nn.Module):
 
     Examples
     --------
```
Comment (rbharath) [inline deepchem/models/torch_models/layers.py:27]: I'm assuming these changes will be rebased out once the MLP PR is merged?

File: deepchem/models/torch_models/infograph.py [hunk 19]
```diff
@@ -0,0 +1,429 @@
...
+        res = torch.mm(l_enc, g_enc.t())
+
+        E_pos = self.get_positive_expectation(res * pos_mask)
+        E_pos = (E_pos * pos_mask).sum() / pos_mask.sum()
+        E_neg = self.get_negative_expectation(res * neg_mask)
+        E_neg = (E_neg * neg_mask).sum() / neg_mask.sum()
+
+        return E_neg - E_pos
+
+    def global_global_loss_(self, g_enc, g_enc1):
```
Comment (rbharath) [inline deepchem/models/torch_models/infograph.py:251]: Should be `_global_global_loss`?

**Review by rbharath — COMMENTED** [review]
Almost at the finish line here! Some improvements needed on testing and documentation and a rebase after the merge in the MLP layer PR but I think one more pass and we can get this in

File: deepchem/models/torch_models/infograph.py (hunk 12 above)
Comment (tonydavis629) [reply deepchem/models/torch_models/infograph.py:116]: It is necessary for all torch_models so they can access the optional torch_model arguments (like batch_size, tensorboard, etc)

File: deepchem/models/torch_models/infograph.py (hunk 15 above)
Comment (tonydavis629) [reply deepchem/models/torch_models/infograph.py:112]: Infograph is written with different types of training already in mind, whereas I think many models are made with only one type of training task. Modifying the loss_func is how I intend the training task to be modified in ModularTorchModel. I think in the case of infograph it makes sense, but maybe it should not be the usual pattern.

File: docs/source/api_reference/models.rst (hunk 8 above)
Comment (rbharath) [reply docs/source/api_reference/models.rst:402]: Quick reminder on this TODO

**Review by rbharath — APPROVED** [review]
LGTM. One last minor comment about moving the layer documentation to `layers.rst` but clear to merge after once CI is good
